docs/internals/task_lifecycle.md

```diff
@@ -0,0 +1,186 @@
+---
+title:  "Task Lifecycle"
+nav-title: Task Lifecycle
+nav-parent_id: internals
+nav-pos: 5
+---
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+A task in Flink is the basic unit of execution. It is the place where each parallel instance of an operator is executed. 
+
+This document goes through the different phases in the lifecycle of the `StreamTask`, which is the base for all different 
+task sub-types in Flink's streaming engine. 
+
+* This will be replaced by the TOC
+{:toc}
+
+## Operator Lifecycle in a nutshell
+
+Because the task is the entity that executes your operators, its lifecycle is tightly integrated with that of an operator. 
+So, we will briefly mention the basic methods representing the lifecycle of an operator before diving into those of the 
+`StreamTask` itself. The list is presented below in the order that each of the methods is called. Given that an operator
+can have a user-defined function (*UDF*), below each of the operator methods we also present (indented) the methods in 
+the lifecycle of UDF that it calls. These methods are available if your operator extends the `AbstractUdfStreamOperator`, 
+which is the basic class for all operators that execute UDFs.
+
+        // initialization
+        OPERATOR::setup
+            UDF::setRuntimeContext
+        OPERATOR::initializeState
+        OPERATOR::open
+            UDF::open
+        
+        // processing
+        OPERATOR::processElement
+            UDF::run
+        OPERATOR::processWatermark
+        
+        // checkpointing (called asynchronously)
+        OPERATOR::snapshotState
```

Comment:
Again, this could be called multiple times in the lifecycle. Once for each triggered checkpoint.